_posts/2023-02-01-dask-on-flyte.md

```diff
@@ -0,0 +1,151 @@
+
+
+def inc(x):
+    return x + 1
+
+
+@task(
+    task_config=Dask(
+        scheduler=Scheduler(
+            requests=Resources(cpu="2")
+        ),
+        workers=WorkerGroup(
+            number_of_workers=4,
+            limits=Resources(cpu="8", mem="32Gi")
+        )
+    )
+)
+def increment_numbers(list_length: int) -> List[int]:
+    client = Client()
+    futures = client.map(inc, range(list_length))
+    return client.gather(futures)
+```
+
+This task will run locally (using a regular `distributed.Client()`) and can then scale to arbitrary cluster sizes after is has been [registered with Flyte](https://docs.flyte.org/projects/cookbook/en/latest/auto/larger_apps/larger_apps_deploy.html).
+
+## What is Flyte?
+
+[Flyte](https://flyte.org/) is a Kubernetes native workflow orchestration engine. Originally developed at [Lyft](https://www.lyft.com/), it is now open source ([Github](https://github.com/flyteorg/flyte)) and Graduate Project from the Linux Foundation. Some it's core features that distinguish it from similar tools like [Airflow](https://airflow.apache.org/) or [Argo](https://argoproj.github.io/) are:
+
+- Caching/Memoization of previously executed tasks
+- Kubernetes native
+- Workflows are defined as Python code, not e.g. `yaml`
+- Strong typing between tasks and workflows using `protobuf`
+- Dynamically generated workflow DAGs at runtime
+- Workflows can be executed locally
+
+A simple workflow would look something like the following:
+
+```python
+from typing import List
+
+import pandas as pd
+
+from flytekit import task, workflow, Resources
+from flytekitplugins.dask import Dask, WorkerGroup, Scheduler
+
+
+@task(
+    task_config=Dask(
+        scheduler=Scheduler(
+            requests=Resources(cpu="2")
+        ),
+        workers=WorkerGroup(
+            number_of_workers=4,
+            limits=Resources(cpu="8", mem="32Gi")
+        )
+    )
+)
+def expensive_data_preparation(input_files: List[str]) -> pd.DataFrame:
+    # Expensive, highly parallel `dask` code
+    ...
+    return pd.DataFrame(...)  # Some large DataFrame, Flyte will handle serialization
+
+
+@task
+def train(input_data: pd.DataFrame) -> str:
+    # Model training, can also use GPU, etc.
+    ...
+    return "s3://path-to-model"
+
+
+@workflow
+def train_model(input_files: List[str]) -> str:
+    prepared_data = expensive_data_preparation(input_files=input_files)
+    return train(input_data=prepared_data)
+```
+
+In the above, both `expensive_data_preparation()` as well as `train()` would be run in their own Pod(s) in Kubernetes, while the `train_model()` workflow is a DSL which creates a Directed Acyclic Graph (DAG) of the workflow. It will determine the order of tasks based on their inputs and outputs. Input and output types (based on the type hints) will be validated at registration time to avoid surprises at runtime.
+
+After registration with Flyte, the workflow can be started from the UI:
```

Comment:
In the example, `expensive_data_preparation()` and `train()` would each run in their own pods within Kubernetes. The `train_model()` workflow, expressed using a domain-specific language (DSL), generates a Directed Acyclic Graph (DAG) that outlines the task order based on inputs and outputs. Input and output types, specified using type hints, are validated during registration to avoid surprises at runtime.